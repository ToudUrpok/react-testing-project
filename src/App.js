import { useEffect, useState } from 'react';
import './App.css';

function App() {
    const [data, setData] = useState('');
    const [toggle, setToggle] = useState(false);
    const [value, setValue] = useState('');

    const onClick = () => setToggle(prev => !prev);

    useEffect(() => {
        setTimeout(() => {
            setData('Async Component');
        }, 1000);
    }, []);

    return (
        <div>
            {toggle && <div data-testid='toggle-elem'>Controlled Component</div>}
            {data && <div style={{color: 'red'}}>{data}</div>}
            <h1>React Testing Library</h1>
            <button data-testid='toggle-btn' onClick={onClick}>Toggle</button>
            <h1 data-testid='input-value'>{value}</h1>
            <input 
                type='text' 
                placeholder='input value' 
                value={value}
                onChange={e => setValue(e.target.value)}
            />
        </div>
    );
}

export default App;
