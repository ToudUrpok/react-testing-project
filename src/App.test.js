import { act, fireEvent, render, screen } from '@testing-library/react';
import App from './App';
import userEvent from '@testing-library/user-event';

describe('test App component', () => {
    test('first test', () => {
        render(<App />);
        const h1 = screen.getByText(/react testing library/i);
        const btn = screen.getByRole('button');
        const input = screen.getByPlaceholderText(/input value/i);
        expect(h1).toBeInTheDocument();
        expect(btn).toBeInTheDocument();
        expect(input).toMatchSnapshot();
    });

    test('not existed component', () => {
        render(<App />);
        const notExists = screen.queryByText(/not existed/i);
        expect(notExists).toBeNull();
    });

    test('find component', async () => {
        render(<App />);
        screen.debug();
        const asyncComponent = await screen.findByText(/async component/i);
        expect(asyncComponent).toBeInTheDocument();
        expect(asyncComponent).toHaveStyle({color: 'red'});
        screen.debug();
    });

    test('click event test', () => {
        render(<App />);
        const btn = screen.getByTestId('toggle-btn');
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
        fireEvent.click(btn);
        expect(screen.getByTestId('toggle-elem')).toBeInTheDocument();
        fireEvent.click(btn);
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
    });

    test('input event test', () => {
        render(<App />);
        expect(screen.getByTestId('input-value')).toContainHTML('');
        const input = screen.getByPlaceholderText(/input value/i);
        fireEvent.input(input, {
            target: {value: 'Hello, bitch!'}
        });
        expect(screen.getByTestId('input-value')).toContainHTML('Hello, bitch!');
    });

    test('input test with userEvent', () => {
        render(<App />);
        expect(screen.getByTestId('input-value')).toContainHTML('');
        const input = screen.getByPlaceholderText(/input value/i);
        act(() => {
            userEvent.type(input, 'Hello, bitch!');
        });
        expect(screen.getByTestId('input-value')).toContainHTML('Hello, bitch!');
    });
});
