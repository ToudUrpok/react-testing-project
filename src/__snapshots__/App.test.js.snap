// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test App component first test 1`] = `
<input
  placeholder="input value"
  type="text"
  value=""
/>
`;
